torch/csrc/autograd/custom_function.cpp

```diff
@@ -332,7 +332,7 @@ optional_variable_list _process_backward_mode_ad(
       var.mutable_grad().reset();
       impl::clear_hooks(var);
       if (auto grad_acc_fn = impl::try_get_grad_accumulator(var)) {
-        auto grad_acc = dynamic_cast<AccumulateGrad*>(grad_acc_fn.get());
+        auto grad_acc = std::dynamic_pointer_cast<AccumulateGrad>(grad_acc_fn);
```

Comment:
```suggestion
        auto grad_acc = std::dynamic_pointer_cast<AccumulateGrad>(std::move(grad_acc_fn));
```
Otherwise, this is slightly less efficient than the previous code since it requires a global lock to increment / decrement the smartpointer reference count.

Reply:
Ah whoops, I misread the STL docs. This isn't in C++ until C++20. whoops.

Reply:
This is still slightly more inefficient than what was there before, any reason why to do so?